### PR TITLE
TypePattern with both Type and name not working in some scenarios

### PR DESCRIPTION
**Pull request by subyssurendran666**
## What it does
https://github.com/eclipse-jdt/eclipse.jdt.core/issues/2741

## Author checklist

- [x] I have thoroughly tested my changes
- [x] The change is following the [coding conventions](https://wiki.eclipse.org/Platform/How_to_Contribute#Coding_Conventions)
- [x] I have signed the [Eclipse Contributor Agreement (ECA)](https://www.eclipse.org/legal/ECA.php)

### CODE REVIEW

**Comment by iloveeclipse** [comment]
It would really like to see **meaningfull** commit messages on every commit according to https://github.com/eclipse-jdt/.github/blob/main/CONTRIBUTING.md#commit-message-recommendations

**Comment by stephan-herrmann** [comment]
> It would really like to see **meaningfull** commit messages on every commit according to https://github.com/eclipse-jdt/.github/blob/main/CONTRIBUTING.md#commit-message-recommendations

I fully agree regarding the **first commit in a PR**, but I wonder if follow-up commits (which will eventually be squashed anyway) are fine with a little less ceremony? Some explanation: yes, of course. But in that case I see little use in repeating the title, and perhaps even the footer can be omitted for secondary commits? @iloveeclipse  WDYT?

Personally, I find it tiring to see the PR message template time and again with no added information except for an issue number.

**Comment by iloveeclipse** [comment]
Trivial commits do not need lengthy message but still should contain reference to the ticket they are supposed to fix or to contribute. Non trivial commits **must** explain the main idea of the change in the commit message.

"added fix" is not a commit message I personally would like to see in **any** valuable open source project, but is of course fully OK for the code in some private project.

For me, every commit should be self contained and self explaining, and that mandates following https://github.com/eclipse-jdt/.github/blob/main/CONTRIBUTING.md#commit-message-recommendations

**Comment by stephan-herrmann** [comment]
> Trivial commits do not need lengthy message but still should contain reference to the ticket they are supposed to fix or to contribute. Non trivial commits **must** explain the main idea of the change in the commit message.
> 
> "added fix" is not a commit message I personally would like to see in **any** valuable open source project, but is of course fully OK for the code in some private project.
> 
> For me, every commit should be self contained and self explaining, and that mandates following https://github.com/eclipse-jdt/.github/blob/main/CONTRIBUTING.md#commit-message-recommendations

I'm fine concerning all but one small detail from the guide lines: even [very respected committers](https://github.com/eclipse-jdt/eclipse.jdt.core/pull/2551/commits) don't always repeat the ticket title _in every commit_.

**Comment by iloveeclipse** [comment]
> lines: even [very respected committers](https://github.com/eclipse-jdt/eclipse.jdt.core/pull/2551/commits) don't always repeat the ticket title _in every commit_.

The PR in question looks like it is **only** because some other, much more respected committers don't like rebase :)
So all commits after first three will disappear once the review is done.

Regarding ticket numner in the title in particular: as long as commit contains full ticket url it isn't actually needed. I will update guideline.

**Comment by subyssurendran666** [comment]
@iloveeclipse @stephan-herrmann I will add more meaningful commit messages from now on.

File: org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/ASTConverter.java [hunk 1]
```diff
@@ -3638,9 +3638,9 @@ public Pattern convert(org.eclipse.jdt.internal.compiler.ast.TypePattern pattern
 		if(this.ast.apiLevel < AST.JLS22_INTERNAL) {
 			typePattern = new TypePattern(this.ast);
 		} else {
-			if(pattern.local != null && pattern.local.name[0] == '_' && pattern.local.type != null) {
+			if(pattern.local.type != null) {//If there is a type, then SingleVariableDeclaration
```
Comment (mpalat) [inline org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/ASTConverter.java:3641]: code opt:
replace the if statements with 
typePattern = new TypePattern(this.ast, pattern.local.type != null);

**Review by mpalat — APPROVED** [review]
Please make the change mentioned and submit

+1

File: org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/ASTConverter.java (hunk 1 above)
Comment (subyssurendran666) [reply org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/ASTConverter.java]: Done. Thanks for the wonderful suggestion.

**Comment by stephan-herrmann** [comment]
@subyssurendran666  this change modified all converterJclMin*.jar, whereas only converterJclMin22 has changed sources. 
I'd like to ensure that only relevant changes are committed in the future. To help us understand what caused the changes, here please try to answer:
* did you have all these checked out as Java projects in your workspace?
* did you create the jars by running ant build on JCL/build.xml?
* was workspace build enabled at that time?

**Comment by jarthana** [comment]
> Regarding ticket number in the title in particular: as long as commit contains full ticket url it isn't actually needed. I will update guideline.

I have been using the URL of the PR in the commit message. I guess that's OK?

**Comment by iloveeclipse** [comment]
> I have been using the URL of the PR in the commit message. I guess that's OK?

Why not the URL of the actual issue?

**Comment by jarthana** [comment]
> > I have been using the URL of the PR in the commit message. I guess that's OK?
> 
> Why not the URL of the actual issue?

I usually have the issue number on the title. I remember my discussion with Stephan and we seem to conclude this way. I reckon it helps when we have more than one PR for an issue. And sometimes we may not even have an issue.

**Pull request merged**